src/rules-parser.pegjs

```diff
@@ -188,23 +187,43 @@ ParameterList = "(" _ ")" _ {
     return [];
   }
   / "(" _ head:Identifier tail:(_ "," _ Identifier)* _ ")" _ {
-    var result = [head];
+    var result = [ensureLowerCase(head, "Function arguments")];
     for (var i = 0; i < tail.length; i++) {
       result.push(ensureLowerCase(tail[i][3], "Function arguments"));
     }
     return result;
 }
 
-// TODO: Allow for union types, e.g., Type1 | Type2
-// TODO: Allow for generic types, e.g., Type<A, B>
-TypeExpression "type" =
-  head:Identifier tail:(_ "|" _ id:Identifier { return id; } )* _ {
-    var result = [ensureUpperCase(head, "Type names")];
+TypeExpression  = head:SingleType tail:(_ "|" _ type:SingleType { return type; } )* _ {
+  var result = [head];
+  if (tail.length == 0) {
+    return head;
+  }
+  for (var i = 0; i < tail.length; i++) {
+    result.push(tail[i]);
+  }
+  return ast.unionType(result);
+}
+
+// Type, Type[], or Type<X, ... >
+// Type[] === Map<String, Type>
+SingleType = type:Identifier opt:(isMap:("\[\]") / "<" _ types:TypeList _ ">")? _ {
```

Comment:
Exactly, you read my mind. Retain the typeType but shift out the syntax changes to a different PR. Thanks!